eng_cryptodev.c: add support for ECB and CTR modes via OCF engine.

signed-off-by: Bill Tompkins (Broadcom)

--- openssl-1.0.1c/crypto/engine/eng_cryptodev.c.orig	2012-02-27 11:33:25.000000000 -0500
+++ openssl-1.0.1c/crypto/engine/eng_cryptodev.c	2013-10-04 01:00:59.189450474 -0400
@@ -147,6 +147,12 @@
 	{ CRYPTO_AES_CBC,		NID_aes_128_cbc,	16,	16, },
 	{ CRYPTO_AES_CBC,		NID_aes_192_cbc,	16,	24, },
 	{ CRYPTO_AES_CBC,		NID_aes_256_cbc,	16,	32, },
+	{ CRYPTO_AES_ECB,		NID_aes_128_ecb,	16,	16, },
+	{ CRYPTO_AES_ECB,		NID_aes_192_ecb,	16,	24, },
+	{ CRYPTO_AES_ECB,		NID_aes_256_ecb,	16,	32, },
+	{ CRYPTO_AES_CTR,		NID_aes_128_ctr,	16,	16, },
+	{ CRYPTO_AES_CTR,		NID_aes_192_ctr,	16,	24, },
+	{ CRYPTO_AES_CTR,		NID_aes_256_ctr,	16,	32, },
 	{ CRYPTO_BLF_CBC,		NID_bf_cbc,		8,	16, },
 	{ CRYPTO_CAST_CBC,		NID_cast5_cbc,		8,	16, },
 	{ CRYPTO_SKIPJACK_CBC,		NID_undef,		0,	 0, },
@@ -600,6 +606,85 @@
 	NULL
 };
 
+const EVP_CIPHER cryptodev_aes_ecb = {
+	NID_aes_128_ecb,
+	16, 16, 16,
+	EVP_CIPH_ECB_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+const EVP_CIPHER cryptodev_aes_192_ecb = {
+	NID_aes_192_ecb,
+	16, 24, 16,
+	EVP_CIPH_ECB_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+const EVP_CIPHER cryptodev_aes_256_ecb = {
+	NID_aes_256_ecb,
+	16, 32, 16,
+	EVP_CIPH_ECB_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+const EVP_CIPHER cryptodev_aes_ctr = {
+	NID_aes_128_ctr,
+	16, 16, 16,
+	EVP_CIPH_CTR_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+const EVP_CIPHER cryptodev_aes_192_ctr = {
+	NID_aes_192_ctr,
+	16, 24, 16,
+	EVP_CIPH_CTR_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+const EVP_CIPHER cryptodev_aes_256_ctr = {
+	NID_aes_256_ctr,
+	16, 32, 16,
+	EVP_CIPH_CTR_MODE,
+	cryptodev_init_key,
+	cryptodev_cipher,
+	cryptodev_cleanup,
+	sizeof(struct dev_crypto_state),
+	EVP_CIPHER_set_asn1_iv,
+	EVP_CIPHER_get_asn1_iv,
+	NULL
+};
+
+
 /*
  * Registered by the ENGINE when used to find out how to deal with
  * a particular NID in the ENGINE. this says what we'll do at the
@@ -637,10 +722,29 @@
 	case NID_aes_256_cbc:
 		*cipher = &cryptodev_aes_256_cbc;
 		break;
+	case NID_aes_128_ecb:
+		*cipher = &cryptodev_aes_ecb;
+		break;
+	case NID_aes_192_ecb:
+		*cipher = &cryptodev_aes_192_ecb;
+		break;
+	case NID_aes_256_ecb:
+		*cipher = &cryptodev_aes_256_ecb;
+		break;
+	case NID_aes_128_ctr:
+		*cipher = &cryptodev_aes_ctr;
+		break;
+	case NID_aes_192_ctr:
+		*cipher = &cryptodev_aes_192_ctr;
+		break;
+	case NID_aes_256_ctr:
+		*cipher = &cryptodev_aes_256_ctr;
+		break;
 	default:
 		*cipher = NULL;
 		break;
 	}
+
 	return (*cipher != NULL);
 }
 
